README.md
ecca-datingsite
===============

The world's most secure dating site. 

debian/changelog
ecca-dating (0.1) UNRELEASED; urgency=low

  * Initial packaging

 -- Guido Witmond <guido@witmond.nl>  Sun, 31 Oct 2015 20:49:41 +0200


db1.go
// Database connection

package main

import (
	"database/sql"
	_ "github.com/gwenn/gosqlite"

	// utils
	"log"
)

var db *sql.DB

func init() {
	var err error
	log.Printf("opening database file: %v", *dbFile)
	db, err = sql.Open("sqlite3", *dbFile)
	check(err)

	_, err = db.Exec("CREATE TABLE aliens (cn TEXT, race TEXT, occupation TEXT)")
	// check(err) // ignore

	_, err = db.Exec("CREATE TABLE messages (toCN TEXT, fromCN, ciphertext BLOB)")
	// check(err) // ignore
}


// saveAlien inserts or updates an Alien record
func saveAlien(alien Alien) {
	existing := getAlien(alien.CN)
	var result sql.Result
	var err error
	if existing == nil {
		insert, err := db.Prepare("INSERT INTO aliens (cn, race, occupation) values (?, ?, ?)")
		check(err)
		defer insert.Close()

		result, err = insert.Exec(alien.CN, alien.Race, alien.Occupation)
		check(err)
	} else {
		update, err := db.Prepare("UPDATE aliens SET race = ?, occupation = ? WHERE cn = ?")
		check(err)
		defer update.Close()

		result, err = update.Exec(alien.Race, alien.Occupation, alien.CN)
		check(err)
	}
	count, err := result.RowsAffected()
	check(err)
	log.Printf("Inserted %d rows", count)
}

func getAliens() (aliens []Alien) {
	rows, err := db.Query("SELECT cn, race, occupation FROM aliens")
	check(err)
	defer rows.Close()
	for rows.Next() {
		var alien Alien
		rows.Scan(&alien.CN, &alien.Race, &alien.Occupation)
		aliens = append(aliens, alien)
	}
	return
}

// getAlien gets one alien
func getAlien(cn string) (*Alien) {
	rows, err := db.Query("SELECT cn, race, occupation FROM aliens WHERE cn = ?", cn)
	check(err)
	defer rows.Close()
	if rows.Next() {
		var alien Alien
		rows.Scan(&alien.CN, &alien.Race, &alien.Occupation)
		return &alien
	}
	return nil
}



func saveMessage(message Message) {
	insert, err := db.Prepare("INSERT INTO messages (toCN, fromCN, ciphertext) values (?, ?, ?)")
	check(err)
	defer insert.Close()

	result, err := insert.Exec(message.ToCN, message.FromCN, message.Ciphertext)
	check(err)
	count, err := result.RowsAffected()
	check(err)
	log.Printf("Inserted %d rows", count)
}

func getMessages(toCN string) (messages []Message) {
	query := "SELECT toCN, fromCN, ciphertext FROM messages WHERE toCN = ?"
	log.Printf("query is: %s  with %s\n", query, toCN)
	rows, err := db.Query(query, toCN)
	check(err)
	defer rows.Close()
	for rows.Next() {
		var message Message
		rows.Scan(&message.ToCN, &message.FromCN, &message.Ciphertext)
		messages = append(messages, message)
	}
	return
}


dating.go
// Eccentric Authentication Dating Site Example
//
// Show the cryptographic message signing and encryption to provide anonymous
// but secure and private communication.
//
// Copyright 2013, Guido Witmond <guido@witmond.nl>
// Licensed under AGPL v3 or later. See LICENSE.

package main

import (
	"log"
	"net/http"
	"net/url"
	"crypto/tls"
	"html/template"
	"flag"

	"github.com/gwitmond/eccentric-authentication" // package eccentric

	// These are for the data storage
	//"github.com/coopernurse/gorp"
	//"database/sql"
	//_ "github.com/mattn/go-sqlite3"
)


// The things to set before running.
var certDir = flag.String("config", "cert", "Directory where the certificates and keys are found.")
var fpcaCert = flag.String("fpcaCert", "applicationFPCA.cert.pem", "File with the Certificate of the First Party Certificate Authority that we accept for our clients.")
var fpcaURL = flag.String("fpcaURL", "https://register-application.example.nl", "URL of the First Party Certificate Authority where clients can get their certificate.")
var hostname = flag.String("hostname", "application.example.nl", "Hostname of the application. Determines which cert.pem and key.pem are used for the TLS-connection.")
var bindAddress = flag.String("bind", "[::]:443", "Address and port number where to bind the listening socket.")
var dbFile =  flag.String("datastore", "/var/lib/ecca-dating/data/ecca-dating.sqlite3", "Path for the ecca-dating.sqlite3 datastore")

// global state
var ecca = eccentric.Authentication{}

var templates = template.Must(template.ParseFiles(
	"templates/homepage.template",
	"templates/editProfile.template",
	"templates/aliens.template",
	"templates/showAlien.template",
	"templates/readMessage.template",
	"templates/sendMessage.template",
	"templates/needToRegister.template",
	"templates/menu.template"))


func init() {
	http.HandleFunc("/", homePage)
	http.HandleFunc("/aliens", showAliens)
	http.HandleFunc("/show", showProfile)

	http.Handle("/profile", ecca.LoggedInHandler(editProfile, "needToRegister.template"))

	http.Handle("/read-messages", ecca.LoggedInHandler(readMessages, "needToRegister.template"))
	http.Handle("/send-message", ecca.LoggedInHandler(sendMessage, "needToRegister.template"))

	http.Handle("/initiate-direct-connection", ecca.LoggedInHandler(initiateDirectConnection, "needToRegister.template"))

	http.Handle("/static/", http.FileServer(http.Dir(".")))
}


func main() {
	flag.Parse()
	ecca = eccentric.Authentication{
		RegisterURL:  *fpcaURL, // "https://register-dating.wtmnd.nl:10444/register-pubkey",
		Templates: templates,   //Just copy the templates variable
	}

	// This CA-pool specifies which client certificates can log in to our site.
	pool := eccentric.ReadCert( *certDir + "/" + *fpcaCert) // "datingLocalCA.cert.pem"

	log.Printf("Started at %s. Go to https://%s/ + port", *bindAddress, *hostname)

	server := &http.Server{Addr: *bindAddress,
		TLSConfig: &tls.Config{
			ClientCAs: pool,
			ClientAuth: tls.VerifyClientCertIfGiven},
	}
	// Set  the server certificate to encrypt the connection with TLS
	ssl_certificate := *certDir + "/" + *hostname + ".cert.pem"
	ssl_cert_key   := *certDir + "/" + *hostname + ".key.pem"
	check(server.ListenAndServeTLS(ssl_certificate, ssl_cert_key))
}


func homePage(w http.ResponseWriter, req *http.Request) {
	if req.URL.Path == "/" {
		check(templates.ExecuteTemplate(w, "homepage.template",  nil))
		return;
	}
	http.NotFound(w, req)
}


// editProfile lets the user fill in his/her profile data to lure the aliens into the hive.
func editProfile(w http.ResponseWriter, req *http.Request) {
	// LoggedInHander made sure our user is logged in with a correct certificate
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "GET":
		alien := getAlien(cn)  // alien or nil
		check(templates.ExecuteTemplate(w, "editProfile.template", map[string]interface{}{
			"CN": cn,
			"alien": alien,
			"races": races,
			"occupations": occupations,
		}))

	case "POST":
		req.ParseForm()
		saveAlien(Alien{
			CN: cn,
			Race: req.Form.Get("race"),
			Occupation: req.Form.Get("occupation"),
		})
		//TODO: make a nice template with a menu and a redirect-link.
		w.Write([]byte(`<html><p>Thank you for your entry. <a href="/aliens">Show all aliens.</a></p></html>`))

	default:
		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )
	}
}


// Checked sets the checked attribute.
// To be called from within templates.
func (alien *Alien) Checked(data string) string {
 	if alien == nil { return "" } // no data, nothing selected
 	if alien.Race == data { return "checked"} // if the data is in the Alien.Race -> true
 	if alien.Occupation == data { return "checked" } // or if the data is in the Occup. -> true
 	return ""
}


// Show all aliens, no authentication required
func showAliens (w http.ResponseWriter, req *http.Request) {
	aliens := getAliens()
	check(templates.ExecuteTemplate(w, "aliens.template", map[string]interface{}{
		"aliens": aliens,
		"races": races,
		"occupations": occupations,
	}))
}

// Show all aliens, no authentication required
func showProfile (w http.ResponseWriter, req *http.Request) {
	req.ParseForm()
	cn := req.Form.Get("alien")
	alien := getAlien(cn)
	if alien == nil {
		http.NotFound(w, req)
		return
	}
	check(templates.ExecuteTemplate(w, "showAlien.template", map[string]interface{}{
		"alien": alien,
		"IdURL": IdURL(cn),
	}))
}


func IdURL (cn string) *url.URL {
	url, err := url.Parse(*fpcaURL)
	url.Path = "/get-certificate"
	check(err)
	q := url.Query()
	q.Set("nickname", cn)
	url.RawQuery = q.Encode()
	return url
}


// readMessages shows you the messages other aliens have sent you.
func readMessages (w http.ResponseWriter, req *http.Request) {
	// User is logged in
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "GET":
		// set this header to signal the user's Agent to perform data decryption.
		w.Header().Set("Eccentric-Authentication", "decryption=\"required\"")
		w.Header().Set("Content-Type", "text/html, charset=utf8")
		messages := getMessages(cn)
		check(templates.ExecuteTemplate(w, "readMessage.template", map[string]interface{}{
			"CN": cn,
			"messages": messages,
		}))

	default:
 		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )

	}
}


// sendMessage takes an encrypted message and delivers it at the message box of the recipient
// Right now, that's our own dating site. It could perform a MitM.
// See: http://eccentric-authentication.org/eccentric-authentication/private_messaging.html
func sendMessage(w http.ResponseWriter, req *http.Request) {
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "GET":
		req.ParseForm()
		toCN := req.Form.Get("addressee")
		idURL := IdURL(toCN)
 		check(templates.ExecuteTemplate(w, "sendMessage.template", map[string]interface{}{
			"CN": cn,            // from us
			"ToCN": toCN,        // to recipient
			"IdURL": idURL,      // where to find the certificate with public key
		}))

	case "POST":
		req.ParseForm()
		ciphertext := req.Form.Get("ciphertext")
		if ciphertext == "" {
			w.Write([]byte(`<html><p>Your message was not encrypted. We won't accept it. Please use the ecca-proxy.</p></html>`))
			return
		}
		saveMessage(Message{
			FromCN: cn,
			ToCN: req.Form.Get("addressee"),
			Ciphertext: ciphertext,
		})
		w.Write([]byte(`<html><p>Thank you, your message will be delivered at galactic speed.</p></html>`))

	default:
 		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )
	}

}


// initiateDirectConnection instructs the ecca-proxy to
// - set up a listener,
// - invite the other via a private message
// - and return a report to the user requesting the connection.
func initiateDirectConnection(w http.ResponseWriter, req *http.Request) {
	// get the inviter
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "POST":
		req.ParseForm()

		// ciphertext is the invitation made by the proxy that intercepted our request.
		// It contains the invitation to toCN, store it and pray they call back.
		ciphertext := req.Form.Get("ciphertext")
		if ciphertext == "" {
			w.Write([]byte(`<html><p>Your message was not encrypted. We won't accept it. Please use the ecca-proxy.</p></html>`))
			return
		}
		saveMessage(Message{
			FromCN: cn,
			ToCN: req.Form.Get("addressee"),
			Ciphertext: ciphertext,
		})
		// check(templates.ExecuteTemplate(w, "sendMessage.template", map[string]interface{}{
		//      "CN": cn,            // from us
		// 	"ToCN": toCN,        // to recipient
		// 	"IdURL": idURL,      // where to find the certificate with public key
		// }))
		w.Write([]byte(`<html><p> The big wait....etc.<br>Expect report here.</p></html>`))

	default:
 		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )
	}
}



func check(err error) {
	if err != nil {
		panic(err)
	}
}


// Example data

var races = map[string]string{
	"a-mars": "Man from Mars",
	"b-venus": "Vamp from Venus",
	"c-creature": "A small furry creature from Alpha Centauri",
}

var occupations = map[string]string{
	"a-captain": "Captain of a Death Star",
	"b-barkeeper": "Barkeeper at a deep space station",
	"c-redshirt": "Redshirt at an away mission",
}


// Marshalling

type Alien struct {
	CN string
	Race string
	Occupation string
}

// get full description for race-id
func (alien Alien) GetRace() (string) {
	return races[alien.Race]
}

// get full description for occupation-id
func (alien Alien) GetOccupation() (string) {
	return occupations[alien.Occupation]
}

// Message struct is used to display received messages
type Message struct {
	ToCN string
	FromCN string
	Ciphertext string // []byte  // don't convert to utf-8 string and back
}




util/make-ca-certs.go
package main

import (
        "crypto/rand"
        "crypto/rsa"
        "crypto/tls"
        "crypto/x509"
        "crypto/x509/pkix"
        "encoding/pem"
        "errors"
        "io"
	//"io/ioutil"
	"os"
        "math/big"
        //"net"
        //"path/filepath"
        //"syscall"
        "time"
	//"log"
)


func main() {
	// Generate a self signed CA cert & key.
	caCert, caKey, err := generateCA("The World's most secure dating site")
	handle(err)
	writePair("datingTestRootCA", caCert, caKey)

	// Generate the Key and certificate that sign the client certificates
	localCaCert, localCaKey, err := generateLocalCA("Register at the World's most secure dating site", caCert, caKey)
	handle(err)
	writePair("datingTestLocalCA", localCaCert, localCaKey)
	
        // Generate an alpha cert signed by our CA cert
        alphaCert, alphaKey, err := generateCert("datingtest.wtmnd.nl", caCert, caKey)
        handle(err)
	writePair("datingtest.wtmnd.nl", alphaCert, alphaKey)

        // Generate an beta cert signed by our CA cert
        betaCert, betaKey, err := generateCert("register-datingtest.wtmnd.nl", caCert, caKey)
        handle(err)
	writePair("register-datingtest.wtmnd.nl", betaCert, betaKey)
}

func writePair(serverName string, cert *x509.Certificate, key *rsa.PrivateKey) {
	cBytes := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})
	kBytes := pem.EncodeToMemory(&pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(key)})	
	err := writeFile(serverName + ".cert.pem", cBytes, 0444)
	handle(err)
	err = writeFile(serverName + ".key.pem", kBytes, 0400)
	handle(err)
}

// writeFile writes data to a file named by filename.
// If the file does not exist, WriteFile creates it with permissions perm;
// It does not overwrite files.
func writeFile(filename string, data []byte, perm os.FileMode) error {
        f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_EXCL, perm)
        if err != nil {
                return err
        }
        n, err := f.Write(data)
        f.Close()
        if err == nil && n < len(data) {
                err = io.ErrShortWrite
        }
        return err
}

func generateCA(serverName string) (*x509.Certificate, *rsa.PrivateKey, error) {
	priv, err := rsa.GenerateKey(rand.Reader, 4096)
	if err != nil {
		return nil, nil, err
	}

	serial := randBigInt()
	keyId := randBytes()

	template := x509.Certificate{
		Subject: pkix.Name{
			CommonName: serverName,
		},

		SerialNumber:   serial,
		SubjectKeyId:   keyId,
		AuthorityKeyId: keyId,
		NotBefore:      time.Now().Add(-5 * time.Minute).UTC(),
		NotAfter:       time.Now().AddDate(5, 0, 0).UTC(),

		KeyUsage:              x509.KeyUsageCertSign,
		BasicConstraintsValid: true,
		IsCA:                  true,
	}

	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &priv.PublicKey, priv)
	if err != nil {
		return nil, nil, err
	}

	certs, err := x509.ParseCertificates(derBytes)
	if err != nil {
		return nil, nil, err
	}

	if len(certs) != 1 {
		return nil, nil, errors.New("Failed to generate a parsable certificate")
	}

	return certs[0], priv, nil
}


func generateLocalCA(localCaName string, caCert *x509.Certificate, caKey *rsa.PrivateKey) (*x509.Certificate, *rsa.PrivateKey, error) {
	priv, err := rsa.GenerateKey(rand.Reader, 3072)
	if err != nil {
		return nil, nil, err
	}
	
	serial := randBigInt()
	keyId := randBytes()

	template := x509.Certificate{
		Subject: pkix.Name{
			CommonName: localCaName,
		},

		SerialNumber:   serial,
		SubjectKeyId:   keyId,
		AuthorityKeyId: caCert.AuthorityKeyId,
		NotBefore:      time.Now().Add(-5 * time.Minute).UTC(),
		NotAfter:       time.Now().AddDate(5, 0, 0).UTC(),

		KeyUsage:              x509.KeyUsageCertSign,
		BasicConstraintsValid: true,
		IsCA:                  true,
	}

	derBytes, err := x509.CreateCertificate(rand.Reader, &template, caCert, &priv.PublicKey, caKey)
	if err != nil {
		return nil, nil, err
	}
	certs, err := x509.ParseCertificates(derBytes)
	if err != nil {
		return nil, nil, err
	}
	if len(certs) != 1 {
		return nil, nil, errors.New("Failed to generate a parsable certificate")
	}

	return certs[0], priv, nil
}

func generateCert(serverName string, caCert *x509.Certificate, caKey *rsa.PrivateKey) (*x509.Certificate, *rsa.PrivateKey, error) {
	priv, err := rsa.GenerateKey(rand.Reader, 1024)
	if err != nil {
		return nil, nil, err
	}

	serial := randBigInt()
	keyId := randBytes()

	template := x509.Certificate{
		Subject: pkix.Name{
			CommonName: serverName,
		},

		SerialNumber:   serial,
		SubjectKeyId:   keyId,
		AuthorityKeyId: caCert.AuthorityKeyId,
		NotBefore:      time.Now().Add(-5 * time.Minute).UTC(),
		NotAfter:       time.Now().AddDate(2, 0, 0).UTC(),
	}

	derBytes, err := x509.CreateCertificate(rand.Reader, &template, caCert, &priv.PublicKey, caKey)
	if err != nil {
		return nil, nil, err
	}
	certs, err := x509.ParseCertificates(derBytes)
	if err != nil {
		return nil, nil, err
	}
	if len(certs) != 1 {
		return nil, nil, errors.New("Failed to generate a parsable certificate")
	}

	return certs[0], priv, nil
}


func generatePair(serverName string, caCert *x509.Certificate, caKey *rsa.PrivateKey) (tls.Certificate, error) {
	cert, key, err := generateCert(serverName, caCert, caKey)
		
	if err != nil {
		return tls.Certificate{}, err
			
		}
	return x509Pair(cert, key)
}

func x509Pair(cert *x509.Certificate, key *rsa.PrivateKey) (tls.Certificate, error) {
	cBytes := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw})
	kBytes := pem.EncodeToMemory(&pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(key)})
	
	return tls.X509KeyPair(cBytes, kBytes)
}

var (
        maxInt64 int64 = 0x7FFFFFFFFFFFFFFF
        maxBig64       = big.NewInt(maxInt64)
)


func randBigInt() (value *big.Int) {
	value, _ = rand.Int(rand.Reader, maxBig64)
	return
}

func randBytes() (bytes []byte) {
	bytes = make([]byte, 20)
	rand.Read(bytes)
	return
}

func handle(err error) {
	if err != nil {
		panic(err.Error())
	}
}


main.go
// Eccentric Authentication Dating Site Example
//
// Show the cryptographic message signing and encryption to provide anonymous
// but secure and private communication.
//
// Copyright 2013, Guido Witmond <guido@witmond.nl>
// Licensed under AGPL v3 or later. See LICENSE.

package main

import (
	"log"
	"net/http"
	"net/url"
	"crypto/tls"
	"html/template"
	"flag"

	"github.com/gwitmond/eccentric-authentication" // package eccentric

	"database/sql"
	_ "github.com/gwenn/gosqlite"
)


// The things to set before running.
var certDir = flag.String("cert", "cert", "Directory where the certificates and keys are found.") 
var fpcaCert = flag.String("fpcaCert", "applicationFPCA.cert.pem", "File with the Certificate of the First Party Certificate Authority that we accept for our clients.")
var fpcaURL = flag.String("fpcaUrl", "https://register-application.example.nl", "URL of the First Party Certificate Authority where clients can get their certificate.")
var hostname = flag.String("hostname", "application.example.nl", "Hostname of the application. Determines which cert.pem and key.pem are used for the TLS-connection.")
var bindAddress = flag.String("bind", "[::]:443", "Address and port number where to bind the listening socket.") 
//var namespace = flag.String("namespace", "", "Name space that we are signing. I.E. <cn>@@example.com. Specifiy the part after the @@.")

var ecca= eccentric.Authentication{
	RegisterURL:  *fpcaURL, // "https://register-dating.wtmnd.nl:10444/register-pubkey",
	Templates: templates,   //Just copy the templates variable
}

var templates = template.Must(template.ParseFiles(
	"templates/homepage.template",
	"templates/editProfile.template",
	"templates/aliens.template",
	"templates/readMessage.template",
	"templates/sendMessage.template",
	"templates/needToRegister.template",
	"templates/menu.template",
	"templates/tracking.template")) 


func init() {
	http.HandleFunc("/", homePage)
	http.HandleFunc("/aliens", showProfiles)

	http.Handle("/profile", ecca.LoggedInHandler(editProfile, "needToRegister.template"))

	http.Handle("/read-messages", ecca.LoggedInHandler(readMessages, "needToRegister.template"))
	http.Handle("/send-message", ecca.LoggedInHandler(sendMessage, "needToRegister.template"))

	http.Handle("/static/", http.FileServer(http.Dir(".")))
}


func main() {
	flag.Parse()
	// This CA-pool specifies which client certificates can log in to our site.
	pool := eccentric.ReadCert( *certDir + "/" + *fpcaCert) // "datingLocalCA.cert.pem"
	
	log.Printf("Started at %s. Go to https://%s/ + port", *bindAddress, *hostname)
	
	server := &http.Server{Addr: *bindAddress,
		TLSConfig: &tls.Config{
			ClientCAs: pool,
			ClientAuth: tls.VerifyClientCertIfGiven},
	}
	// Set  the server certificate to encrypt the connection with TLS
	ssl_certificate := *certDir + "/" + *hostname + ".cert.pem"
	ssl_cert_key   := *certDir + "/" + *hostname + ".key.pem"
	
	check(server.ListenAndServeTLS(ssl_certificate, ssl_cert_key))
}


func homePage(w http.ResponseWriter, req *http.Request) {
	if req.URL.Path == "/" {
		check(templates.ExecuteTemplate(w, "homepage.template",  nil))
	}
	http.NotFound(w, req)
}


// editProfile lets the user fill in his/her profile data to lure the aliens into the hive.
func editProfile(w http.ResponseWriter, req *http.Request) {
	// LoggedInHander made sure our user is logged in.
	// If not, this will give a nice 500 Internal Server Error.
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "GET": 
		alien := getAlien(cn)  // alien or nil
		check(templates.ExecuteTemplate(w, "editProfile.template", map[string]interface{}{
			"CN": cn,
			"alien": alien,
			"races": races,
			"occupations": occupations,
		}))

	case "POST":
		req.ParseForm()
		saveAlien(Alien{
			CN: cn,
			Race: req.Form.Get("race"),
			Occupation: req.Form.Get("occupation"),
		})
		//TODO: make a nice template with a menu and a redirect-link.
		w.Write([]byte(`<html><p>Thank you for your entry. <a href="/aliens">Show all aliens.</a></p></html>`))

	default: 
		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )
	}
}


// Checked sets the checked attribute.
// To be called from within templates.
func (alien *Alien) Checked(data string) string {
 	if alien == nil { return "" } // no data, nothing selected
 	if alien.Race == data { return "checked"} // if the data is in the Alien.Race -> true
 	if alien.Occupation == data { return "checked" } // or if the data is in the Occup. -> true
 	return ""
}


// Show profiles, no authentication required
func showProfiles (w http.ResponseWriter, req *http.Request) {
	aliens := getAliens()
	check(templates.ExecuteTemplate(w, "aliens.template", map[string]interface{}{
		"aliens": aliens,
		"races": races,
		"occupations": occupations,
	}))
}


// readMessages shows you the messages other aliens have sent you.
func readMessages (w http.ResponseWriter, req *http.Request) {
	// User is logged in
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "GET": 
		// set this header to signal the user's Agent to perform data decryption.
		w.Header().Set("Eccentric-Authentication", "decryption=\"required\"")
		w.Header().Set("Content-Type", "text/html, charset=utf8")
		messages := getMessages(cn)
		check(templates.ExecuteTemplate(w, "readMessage.template", map[string]interface{}{
			"CN": cn,
			"messages": messages,
		}))
		
	default:
 		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )

	}
}


// sendMessage takes an encrypted message and delivers it at the message box of the recipient
// Right now, that's our own dating site. It could perform a MitM.
// See: http://eccentric-authentication.org/eccentric-authentication/private_messaging.html
func sendMessage(w http.ResponseWriter, req *http.Request) {
	cn := req.TLS.PeerCertificates[0].Subject.CommonName
	switch req.Method {
	case "GET": 
		req.ParseForm()
		toCN := req.Form.Get("addressee")
		toURL, err := url.Parse("https://register-dating.wtmnd.nl:10444/get-certificate")
		check(err)
		q := toURL.Query()
		q.Set("nickname", toCN)
		toURL.RawQuery = q.Encode()
 		check(templates.ExecuteTemplate(w, "sendMessage.template", map[string]interface{}{
			"CN": cn,
			"ToCN": toCN,
			"ToURL": toURL,
		}))


	case "POST":
		req.ParseForm()
		ciphertext := req.Form.Get("ciphertext")
		if ciphertext == "" {
			w.Write([]byte(`<html><p>Your message was not encrypted. We won't accept it. Please use the ecca-proxy.</p></html>`))
			return
		}
		saveMessage(Message{
			FromCN: cn,
			ToCN: req.Form.Get("addressee"),
			Ciphertext: ciphertext,
		})
		w.Write([]byte(`<html><p>Thank you, your message will be delivered at galactic speed.</p></html>`))

	default:
 		http.Error(w, "Unexpected method", http.StatusMethodNotAllowed )
	}

}



	
func check(err error) {
	if err != nil {
		panic(err)
	}
}


// Example data

var races = map[string]string{
	"a-mars": "Man from Mars",
	"b-venus": "Vamp from Venus",
	"c-creature": "A small furry creature from Alpha Centauri",
}

var occupations = map[string]string{
	"a-captain": "Captain of a Death Star",
	"b-barkeeper": "Barkeeper at a deep space station",
	"c-redshirt": "Redshirt at an away mission",
}


// Marshalling

type Alien struct {
	CN string
	Race string
	Occupation string
}

// get full description for race-id
func (alien Alien) GetRace() (string) {
	return races[alien.Race]
}

// get full description for occupation-id
func (alien Alien) GetOccupation() (string) {
	return occupations[alien.Occupation]
}

// Message struct is used to display received messages
type Message struct {
	ToCN string
	FromCN string
	Ciphertext string // []byte  // don't convert to utf-8 string and back
}

// Database connection

var dbFile = "datingdb.db"
var db *sql.DB
	
func init() {
	var err error
	db, err = sql.Open("sqlite3", dbFile)
	check(err)	
	
	_, err = db.Exec("CREATE TABLE aliens (cn TEXT, race TEXT, occupation TEXT)")
	// check(err) // ignore

	_, err = db.Exec("CREATE TABLE messages (toCN TEXT, fromCN, ciphertext BLOB)")
	// check(err) // ignore
}

// saveAlien inserts or updates an Alien record
func saveAlien(alien Alien) {
	existing := getAlien(alien.CN)
	var result sql.Result
	var err error
	if existing == nil {
		insert, err := db.Prepare("INSERT INTO aliens (cn, race, occupation) values (?, ?, ?)")
		check(err)
		defer insert.Close()
		
		result, err = insert.Exec(alien.CN, alien.Race, alien.Occupation)
		check(err)
	} else {
		update, err := db.Prepare("UPDATE aliens SET race = ?, occupation = ? WHERE cn = ?")
		check(err)
		defer update.Close()

		result, err = update.Exec(alien.Race, alien.Occupation, alien.CN)
		check(err)
	}
	count, err := result.RowsAffected()
	check(err)
	log.Printf("Inserted %d rows", count)
}

func getAliens() (aliens []Alien) {
	rows, err := db.Query("SELECT cn, race, occupation FROM aliens")
	check(err)
	defer rows.Close()
	for rows.Next() {
		var alien Alien
		rows.Scan(&alien.CN, &alien.Race, &alien.Occupation)
		aliens = append(aliens, alien)
	}
	return
}

// getAlien gets one alien
func getAlien(cn string) (*Alien) {
	rows, err := db.Query("SELECT cn, race, occupation FROM aliens WHERE cn = ?", cn)
	check(err)
	defer rows.Close()
	if rows.Next() {
		var alien Alien
		rows.Scan(&alien.CN, &alien.Race, &alien.Occupation)
		return &alien
	}
	return nil
}



func saveMessage(message Message) {
	insert, err := db.Prepare("INSERT INTO messages (toCN, fromCN, ciphertext) values (?, ?, ?)")
	check(err)
	defer insert.Close()

	result, err := insert.Exec(message.ToCN, message.FromCN, message.Ciphertext)
	check(err)
	count, err := result.RowsAffected()
	check(err)
	log.Printf("Inserted %d rows", count)
}

func getMessages(toCN string) (messages []Message) {
	rows, err := db.Query("SELECT toCN, fromCN, ciphertext FROM messages WHERE toCN = ?", toCN)
	check(err)
	defer rows.Close()
	for rows.Next() {
		var message Message
		rows.Scan(&message.ToCN, &message.FromCN, &message.Ciphertext)
		messages = append(messages, message)
	}
	return
}
